crates/orchestrator/src/lib.rs

```diff
@@ -179,6 +207,57 @@ where
         Ok(self.config.clone())
     }
 
+    #[allow(clippy::cast_possible_truncation)]
+    fn register_public_key(
+        &mut self,
+        node_index: u64,
+        pubkey: &mut Vec<u8>,
+    ) -> Result<(), ServerError> {
+        if self.pub_posted.contains(&node_index) {
+            return Err(ServerError {
+                status: tide_disco::StatusCode::BadRequest,
+                message: "Node has already posted public key".to_string(),
+            });
+        }
+        self.pub_posted.insert(node_index);
+
+        // Sishan NOTE: let me know if there's a better way to remove the first extra 8 bytes
+        pubkey.drain(..8);
+        let register_pub_key = <KEY as SignatureKey>::from_bytes(pubkey).unwrap();
+        let register_pub_key_with_stake = register_pub_key.get_stake_table_entry(1u64);
+        self.config.config.known_nodes_with_stake[node_index as usize] =
+            register_pub_key_with_stake;
+        self.nodes_with_pubkey += 1;
+        println!(
+            "Node {:?} posted public key, now total num posted public key: {:?}",
+            node_index, self.nodes_with_pubkey
+        );
+        if self.nodes_with_pubkey >= (self.config.config.total_nodes.get() as u64) {
+            self.peer_pub_ready = true;
+        }
+        Ok(())
+    }
+
+    fn peer_pub_ready(&self) -> Result<bool, ServerError> {
+        if !self.peer_pub_ready {
+            return Err(ServerError {
+                status: tide_disco::StatusCode::BadRequest,
+                message: "Peer's public configs are not ready".to_string(),
+            });
+        }
+        Ok(self.peer_pub_ready)
+    }
+
+    fn get_config_after_peer_collected(&self) -> Result<NetworkConfig<KEY, ELECTION>, ServerError> {
+        if !self.peer_pub_ready {
+            return Err(ServerError {
+                status: tide_disco::StatusCode::BadRequest,
+                message: "Peer's public configs are not ready".to_string(),
+            });
+        }
+        Ok(self.config.clone())
+    }
```

Comment:
I'm thinking maybe we can have a `get_peers_and_config` endpoint to wait for the peers to be ready and then get the config, in case we forget to call `peer_pub_ready` first. But I'm not sure how useful this will be, so curious about your thoughts. 😄

Reply:
Aha, this is exactly what `post_and_wait_all_public_keys()` in `config.rs` does. Here for the function directly updating or posting to orchestrator state, I'm more inclined to keep them atomic and do not use async there.